hack/windows/Install-OVS.ps1

```diff
@@ -244,49 +423,197 @@ function InstallDependency() {
+    # PID and conf.db files to $OVSInstallDir on Windows Node when it is running.
+    icacls $OVSInstallPath /grant "NT AUTHORITY\SYSTEM:(OI)(CI)F" /T
 }
 
-Log "Installation log location: $InstallLog"
+function InstallOVSServices() {
+    param (
+        [parameter(Mandatory = $true)] [string] $OVSInstallPath
+    )
+
+    # Remove the existing OVS Services to avoid issues.
+    If (ServiceExists("ovs-vswitchd")) {
+        stop-service ovs-vswitchd
+        sc.exe delete ovs-vswitchd
+    }
+    if (ServiceExists("ovsdb-server")) {
+        stop-service ovsdb-server
+        sc.exe delete ovsdb-server
+    }
+
+    # Install OpenSSL dependencies.
+    $OVSBinPaths="${OVSInstallPath}\usr\bin;${OVSInstallPath}\usr\sbin"
+    InstallOpenSSLFiles "$OVSBinPaths"
+
+    # Install OVS Services and configure OVSDB.
+    ConfigOVS($OVSInstallPath)
+}
+
+function PrepareOVSLocalFiles() {
+    $OVSDownloadURL = $DefaultOVSDownloadURL
+    $desiredOVSPublishedHash = $DefaultOVSPublishedHash
+    if ($LocalFile -ne "") {
+        if (-not (Test-Path $LocalFile)){
+            Log "Path $LocalFile doesn't exist, exit"
+            exit 1
+        }
 
-CheckIfOVSInstalled
+        $ovsFile = Get-Item $LocalFile
+        if ($ovsFile -is [System.IO.DirectoryInfo])  {
+            return $ovsFile.FullName
+        }
+
+        # $ovsFile as a zip file is supported
+        $attributes = $ovsFile.Attributes
+        if (("$attributes" -eq "Archive") -and ($ovsFile.Extension -eq ".zip" ) ) {
+            $OVSZip = $LocalFile
+            $OVSDownloadURL = ""
+            $OVSPublishedHash = ""
+        } else {
+            Log "Unsupported local file $LocalFile; it should be a zip archive"
+            exit 1
+        }
+    } else {
+        $OVSZip = "$WorkDir\ovs-win64.zip"
+        if ($DownloadURL -ne "" -and $DownloadURL -ne "$OVSDownloadURL") {
+            $OVSDownloadURL = $DownloadURL
+            $desiredOVSPublishedHash = "invalid"
+        }
+    }
+
+    # Extract zip file to $OVSInstallDir
+    if (Test-Path -Path $OVSInstallDir) {
+        Log "$OVSInstallDir already exists, exit OVS installation."
+        exit 1
+    }
+    $removeZipFile = $false
+    if ($OVSDownloadURL -ne "") {
+        DownloadOVS -localZipFile $OVSZip -downloadURL $OVSDownloadURL -desiredHash $desiredOVSPublishedHash
+        $removeZipFile = $true
+    }
+    $ovsInstallParentPath = Split-Path -Path $OVSInstallDir -Parent
+    Expand-Archive -Path $OVSZip -DestinationPath $ovsInstallParentPath | Out-Null
+    if ($removeZipFile) {
+        rm $OVSZip
+    }
+    return $OVSInstallDir
+}
 
-DownloadOVS
+function CopyOVSUtilities() {
+    param (
+        [parameter(Mandatory = $true)] [string] $OVSInstallPath
+    )
+    $hostUsrBinPath = "C:\openvswitch\usr\bin"
+    $usrBinPath="${OVSInstallPath}\usr\bin"
+    if ("$usrBinPath" -ne "$hostUsrBinPath") {
+        if (-not $global:updateOVS) {
```

Comment:
I still think the use of this global file is worse than an unconditional file copy (is copying 20 files slow on Windows?), but at least the new flag name makes more sense, so we can keep it as is.

Reply:
Having some test, the time cost to copy openvswitch/usr/bin is about 47ms.

Reply:
That doesn't really sound like an issue for an initContainer?

Reply:
OK, I could remove the if-condition if the cost is acceptable.